File: Hackerrank/3_month_preparation/python/week1.py
# test variables
arr = [1, 4, 10, 18]
s = "oieee"


def plus_minus(arr):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1)
    len_arr = len(arr)
    positive_numbers = 0
    negative_numbers = 0
    for number in arr:
        if number > 0:
            positive_numbers += 1
        elif number < 0:
            negative_numbers += 1

    print(round(positive_numbers / len_arr, 6))
    print(round(negative_numbers / len_arr, 6))
    print(round((len_arr - positive_numbers - negative_numbers) / len_arr, 6))


def mini_max_sum(arr):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1)
    sum = 0
    minimum_number = arr[0]
    maximum_number = arr[0]
    for number in arr:
        sum += number
        if number < minimum_number:
            minimum_number = number
        if number > maximum_number:
            maximum_number = number

    print("{sum - maximum_number} {sum - minimum_number}")


def time_conversion(s):
    # s has a fixed length
    # Time complexity: O(1)
    # Space complexity (ignoring input): O(1)
    s_len = len(s)
    hour = s[0:2]
    am_pm = s[s_len - 2 :]
    if hour == "12":
        if am_pm == "AM":
            hour = "00"
        else:
            hour = "12"
    elif am_pm == "PM":
        hour = str(int(hour) + 12)

    return hour + s[2 : (s_len - 2)]


def breaking_the_records(scores):
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(1)
    minimum = scores[0]
    maximum = scores[0]
    records = [0, 0]
    for index in range(1, len(scores)):
        score = scores[index]
        print(score, minimum, maximum)
        if score < minimum:
            minimum = score
            records[0] += 1
        if score > maximum:
            maximum = score
            records[1] += 1

    return records


class CamelCase:
    # n being the length of the words
    # Time complexity: O(n)
    # Space complexity (ignoring input): O(n)

    def __init__(self):
        self.read_input()

    def read_input(self):
        continue_loop = True
        number_of_loops = 0
        while continue_loop & number_of_loops < 1e5:
            try:
                number_of_loops += 1
                line = input()
                line = line.strip().split(";")
                self.read_line(line)
            except EOFError:
                continue_loop = False

    def read_line(self, line):
        if line[0] == "S":
            word = self.split(line[2])
        if line[0] == "C":
            word = self.combine(line[2])
            if line[1] == "M":
                word += "()"
            if line[1] == "C":
                word = word[0].upper() + word[1:]

        # Although word is possibly unbound, that is intentional to give an error
        # and I didn't care to treat error here
        print(word)

    def split(self, word: str):
        if word.endswith("()"):
            word = word[:-2]
        splitted_word = word[0].lower()
        word = word[1:]
        for letter in word:
            if letter.upper() == letter:
                splitted_word += " " + letter.lower()
            else:
                splitted_word += letter

        return splitted_word

    def combine(self, words):
        combined_words = ""
        index = 0
        while index < len(words):
            letter = words[index]
            if letter == " ":
                combined_words += words[index + 1].upper()
                index += 1
            else:
                combined_words += words[index]
            index += 1

        return combined_words

def divisible_sum_pairs(n: int, k: int, ar: list[int]):
    #Time complexity: O(n+k)
    #Space complexity (ignoring input): O(k)
    possible_remainders = {}
    total_pairs = 0

    for number in ar:
        remainder = number % k
        if remainder in possible_remainders:
            possible_remainders[remainder] += 1
        else:
            possible_remainders[remainder] = 1

    # For remainder 0 or k/2, the total pairs will be n choose 2
    if 0 in possible_remainders:
        total_pairs += int(possible_remainders[0] * (possible_remainders[0] - 1) / 2)
    k_is_pair = k % 2 == 0
    half_k = int(k / 2)
    if k_is_pair and (half_k in possible_remainders):
        total_pairs += int(
            possible_remainders[half_k] * (possible_remainders[half_k] - 1) / 2
        )

    # For the rest of the remainders, just need to multiply
    for remainder in range(1, int((k + 1) / 2)):
        if (remainder in possible_remainders) and (
            (k - remainder) in possible_remainders
        ):
            total_pairs += int(
                possible_remainders[remainder] * (possible_remainders[k - remainder])
            )

    return total_pairs

def sparse_arrays(strings, queries):
    #Time complexity: O(n+m)
    #Space complexity (ignoring input): O(n+m)
    strings_dictionary = {}

    for string in strings:
        if string in strings_dictionary:
            strings_dictionary[string] += 1
        else:
            strings_dictionary[string] = 1

    strings_found = []
    for query in queries:
        if query in strings_dictionary:
            strings_found.append(strings_dictionary[query])
        else:
            strings_found.append(0)

    return strings_found

File: Hackerrank/3_month_preparation/rust/src/week1.rs
use std::collections::HashMap;

pub fn plus_minus(arr: &[i32]) {
    let len_arr = arr.len() as f64;
    let mut positive_numbers = 0;
    let mut zero_numbers = 0;
    for number in arr.iter().skip(1) {
        if number > &0 {
            positive_numbers += 1;
        }
        if number == &0 {
            zero_numbers += 1;
        }
    }

    let positive_numbers = positive_numbers as f64;
    let zero_numbers = zero_numbers as f64;

    println!("{:.6}", (positive_numbers / len_arr));
    println!(
        "{:.6}",
        ((len_arr - zero_numbers - positive_numbers) / len_arr)
    );
    println!("{:.6}", (zero_numbers / len_arr));
}

pub fn mini_max_sum(arr: &[i32]) {
    let mut sum = 0_i64;
    let mut minimum_number = arr[0] as i64;
    let mut maximum_number = arr[0] as i64;
    for &number in arr.iter().skip(1) {
        sum += number as i64;
        if (number as i64) < minimum_number {
            minimum_number = number as i64
        }
        if (number as i64) > maximum_number {
            maximum_number = number as i64
        }
    }
    print!("{} {}", sum - maximum_number, sum - minimum_number);
}

pub fn time_conversion(s: &str) -> String {
    let s_len = s.len();
    let hour = &s[0..2];
    let am_pm = &s[s_len - 2..];
    let mut hour_int = hour.parse::<i8>().expect("To be a number");
    if hour == "12" {
        if am_pm == "AM" {
            hour_int = 00;
        }
        if am_pm == "PM" {
            hour_int = 12;
        }
    } else if am_pm == "PM" {
        hour_int += 12;
    }

    format!("{hour_int:02}{}", &s[2..s_len - 2])
}

pub fn breaking_the_records(scores: &[i32]) -> Vec<i32> {
    let mut minimum_score = scores[0];
    let mut maximum_score = scores[0];
    let mut records = vec![0, 0];

    for &score in scores.iter().skip(1) {
        println!("{}", score);
        if score < minimum_score {
            minimum_score = score;
            records[1] += 1;
        }
        if score > maximum_score {
            maximum_score = score;
            records[0] += 1;
        }
    }

    records
}

//Camel case functions
fn split(words: &str) -> String {
    let words = match words.strip_suffix("()") {
        Some(stripped) => stripped,
        None => words,
    };
    let mut chars_iter = words.chars();
    let mut splitted_words = chars_iter
        .next()
        .expect("To have found the first letter")
        .to_lowercase()
        .to_string();
    for letter in chars_iter {
        if letter.is_uppercase() {
            splitted_words.push(' ');
            splitted_words.push_str(&letter.to_lowercase().collect::<String>());
        } else {
            splitted_words.push(letter);
        }
    }

    splitted_words
}

fn combine(words: &str, capitalize: bool) -> String {
    let mut combined_words = if capitalize {
        words
            .chars()
            .next()
            .expect("To have found first letter")
            .to_uppercase()
            .collect::<String>()
    } else {
        words
            .chars()
            .next()
            .expect("To have found first letter")
            .to_string()
    };

    let mut index = 1;
    while index < words.len() {
        let letter = words.chars().nth(index).expect("To have found index");
        if letter == ' ' {
            index += 1;
            combined_words.push_str(
                &words
                    .chars()
                    .nth(index)
                    .expect("To have found index")
                    .to_uppercase()
                    .to_string(),
            )
        } else {
            combined_words.push(letter);
        }
        index += 1;
    }

    combined_words
}
pub fn treat_input(input: String) {
    let lines: Vec<&str> = input.split("\r\n").collect();
    for line in lines {
        let words: Vec<&str> = line.split(";").collect();
        if words[0] == "S" {
            let word = split(words[2]);
            println!("{}", word);
        }
        if words[0] == "C" {
            let capitalize = words[1] == "C";
            let mut word = combine(words[2], capitalize);
            if words[1] == "M" {
                word.push_str("()");
            }
            println!("{}", word);
        }
    }
}

pub fn read_input() {
    let mut input = String::new();
    loop {
        let read_line_result = std::io::stdin().read_line(&mut input);
        match read_line_result {
            Ok(0) => break,
            Ok(_n) => (),
            Err(error) => println!("error: {error}"),
        }
    }
    treat_input(input)
}

pub fn divisible_sum_pairs(n: i32, k: i32, ar: &[i32]) -> i32 {
    //Time complexity: O(n+k)
    //Space complexity (ignoring input): O(k)
    let mut remainder_dictionary = std::collections::HashMap::new();
    for number in ar {
        let remainder = number % k;
        match remainder_dictionary.get(&remainder) {
            Some(n) => remainder_dictionary.insert(remainder, n + 1),
            None => remainder_dictionary.insert(remainder, 1),
        };
    }

    let mut total_pairs = 0;
    //For remainders 0 and k/2, the total pairs will be n choose 2
    println!(
        "->> remainder_dictionary <<- function: divisible_sum_pairs; file: week1.rs\n{:?}",
        remainder_dictionary
    );
    total_pairs += match remainder_dictionary.get(&0) {
        Some(n) => n * (n - 1) / 2,
        None => 0,
    };
    let k_is_pair = k % 2 == 0;
    if k_is_pair {
        total_pairs += match remainder_dictionary.get(&(k / 2)) {
            Some(n) => n * (n - 1) / 2,
            None => 0,
        };
    }

    for remainder in 1..(k + 1) / 2 {
        total_pairs += remainder_dictionary.get(&remainder).unwrap_or(&0)
            * remainder_dictionary.get(&(k - remainder)).unwrap_or(&0);
    }

    total_pairs
}

pub fn sparse_arrays(strings: &[String], queries: &[String]) -> Vec<i32> {
    //Time complexity: O(n+m)
    //Space complexity (ignoring input): O(n+m)
    let mut strings_hash_map = std::collections::HashMap::new();

    for string in strings {
        match strings_hash_map.get(&string) {
            Some(n) => strings_hash_map.insert(string, n + 1),
            None => strings_hash_map.insert(string, 1),
        };
    }

    let mut strings_matched = Vec::new();
    for query in queries {
        strings_matched.push(*strings_hash_map.get(&query).unwrap_or(&0));
    }

    strings_matched
}
